src/certgen_certificate_generator_server.cc

```diff
@@ -280,41 +307,83 @@ namespace certgen
         }
 
         std::string serviceName (params[0]);
-        std::string configFilePath(CONFIG_PATH + serviceName + ".cfg");
-        std::ifstream configFile (configFilePath);
-        if (!configFile)
-        {
-            throw std::runtime_error ("Could not open file " + configFilePath);
-        }
-
-        configFile.exceptions ( std::ifstream::failbit | std::ifstream::badbit );
-        std::stringstream configJson;
-        configJson << configFile.rdbuf();
-        configFile.close ();
-
-        cxxtools::SerializationInfo certgenSi;
-        cxxtools::JsonDeserializer deserializer (configJson);
-        deserializer.deserialize (certgenSi);
-
-        CertificateGeneratorConfig certgenConfig;
-        certgenSi >>= certgenConfig;
+        CertificateGeneratorConfig certgenConfig = generateConfig(serviceName);        
 
         fty::CertificateConfig config = loadConfig (certgenConfig.version(), certgenConfig.certConf());
 
         store (CertificateX509::selfSignSha256(generateKeys(certgenConfig.keyConf()), config), certgenConfig.storageConf());
         return "OK";
     }
 
-    std::string CertificateGeneratorServer::handleGenerateCSR(const fty::Payload & /*params*/)
+    std::string CertificateGeneratorServer::handleGenerateCSR(const fty::Payload & params)
     {
-        std::string empty;
-        return empty;
+        if (params.empty() || params[0].empty ())
+        {
+            throw std::runtime_error ("Missing service name");
+        }
+
+        std::string serviceName (params[0]);
+        CertificateGeneratorConfig certgenConfig = generateConfig(serviceName);        
+
+        fty::CertificateConfig config = loadConfig (certgenConfig.version(), certgenConfig.certConf());
+
+        std::string csrPem;
+
+        // TODO should we implement copy ctor on Keys??
+        m_csrKey.reset(new Keys(generateKeys(certgenConfig.keyConf())));
```

Comment:
m_csrPendings => std::map<ServiceName, CsrX509>

For this you could compare and you do not need to use ptr